### PR TITLE
[FIX]web_editor: mass mail translation broken

### PR DESCRIPTION
**Pull request by dka-odoo**
Issue - https://www.odoo.com/web?#id=727240&view_type=form&model=project.issue&action=963&active_id=681&menu_id=4720

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web_editor/static/src/js/backend.js [hunk 1]
```diff
@@ -388,17 +391,12 @@ var FieldTextHtml = AbstractField.extend({
...
-    //     if (this.lang !== 'en_US' && this.$body.find('.o_dirty').length) {
-    //         this.internal_set_value( this.view.datarecord[this.name] );
-    //          this._dirty_flag = false;
-    //          return this.editor.save();
-    //      } else if (this._dirty_flag && this.editor && this.editor.buildingBlock) {
-    //          this.editor.buildingBlock.clean_for_save();
-    //          this.internal_set_value( this.$content.html() );
-    //      }
-    // },
+    commitChanges: function() {
```
Comment (aab-odoo) [inline addons/web_editor/static/src/js/backend.js]: is this necessary?

Comment (dka-odoo) [inline addons/web_editor/static/src/js/backend.js]: yes, we need to update the translation terms when user changes content in other language (!= en_us)

File: addons/web_editor/static/src/js/backend.js [hunk 2]
```diff
@@ -305,21 +308,21 @@ var FieldTextHtml = AbstractField.extend({
...
-                    self.languages = res;
-                    self.add_button();
-                    setTimeout(self.resize,0);
-                }
+                ]
+            })
+            .then(function(res) {
+                self.languages = res;
+                self.add_button();
+                setTimeout(self.resize,0);
```
Comment (dka-odoo) [inline addons/web_editor/static/src/js/backend.js]: resize is still exist :)

Comment (aab-odoo) [inline addons/web_editor/static/src/js/backend.js]: can you show me where it is defined?

File: addons/web_editor/static/src/js/backend.js (hunk 1 above)
Comment (aab-odoo) [inline addons/web_editor/static/src/js/backend.js]: yeah but i don't understand why you only call setValue if lang == 'en_US' (line 211)

File: addons/web_editor/static/src/js/backend.js [hunk 3]
```diff
@@ -206,7 +206,10 @@ var FieldTextHtml = AbstractField.extend({
         };
         window.odoo[this.callback+"_updown"] = null;
         window.odoo[this.callback+"_downup"] = function (value) {
-            self.set_value(value);
+            //update the field value only if lang = en_US
+            if (self.lang === 'en_US') {
+                self._setValue(value);
+            }
```
Comment (dka-odoo) [inline addons/web_editor/static/src/js/backend.js]: Before saas16, we were using internal_set_value method, which was just updating a value without re rendering, while in new view this method is no more exist and calling _setValue method on each key press, was updating the original value (follow by the write method)
regardless of the language, hence source translation terms was getting broken (when user change a content in other lang)

File: addons/web_editor/static/src/js/backend.js (hunk 2 above)
Comment (aab-odoo) [inline addons/web_editor/static/src/js/backend.js]: ok got it, never mind

File: addons/web_editor/static/src/js/backend.js (hunk 1 above)
Comment (dka-odoo) [inline addons/web_editor/static/src/js/backend.js]: Here, we do not need to update the original value of the field when lang is not en_US,
If user change content in other lang then we simply update the translation terms without touching the source value.

File: addons/web_editor/static/src/xml/backend.xml [hunk 4]
```diff
@@ -22,7 +22,7 @@
     </t>
 
     <t t-name="web_editor.FieldTextHtml.translate">
-        <div t-if="widget.recordData.id &amp;&amp; widget.languages.length" style="margin: 5px; position: fixed; top: 0; left: 0; z-index: 2000;">
+        <div t-if="widget.recordData.id and widget.languages.length and widget.mode === 'edit'" style="margin: 5px; position: fixed; top: 0; left: 0; z-index: 2000;">
```
Comment (Yajo) [inline addons/web_editor/static/src/xml/backend.xml]: This is not valid JS afaik

**Comment by Yajo** [comment]
Is this related to #15622?

**Comment by aab-odoo** [comment]
No, this is only for versions >= saas-16 (after the rewrite of the JS views)

File: addons/web_editor/static/src/xml/backend.xml (hunk 4 above)
Comment (aab-odoo) [inline addons/web_editor/static/src/xml/backend.xml]: qweb handles it

File: addons/web_editor/static/src/js/backend.js [hunk 5]
```diff
@@ -187,6 +187,18 @@ var FieldTextHtmlSimple = basic_fields.DebouncedField.extend({
...
+
+    _renderTranslateButton: function () {
+        if (_t.database.multi_lang && this.field.translate && this.res_id) {
+            $(QWeb.render('web_editor.FieldTextHtml.button.translate', {'widget': this}))
+                .appendTo(this.$('.note-toolbar'))
+                .on('click', this._onTranslate.bind(this));
+        }
+    },
+
+    _onTranslate: function () {
```
Comment (aab-odoo) [inline addons/web_editor/static/src/js/backend.js]: This should be in 'Handlers' section (see https://github.com/odoo/odoo/blob/master/addons/web/static/src/js/fields/abstract_field.js#L395 for example). New coding guidelines: split functions as follows (in that order): first unnamed section containing lifecyle widget functions (init, willStart, start, destroy), then 'Public' section containing all public functions (called by the outside of widget, not prefixed by _), then 'Private' section containing all non event handlers private functions (only called by the widget itself, prefixed by _), and finally 'Handlers' section containing all event handlers (prefixed by _on, as you did). Also document all your functions. Again, you can see AbstractField for example. Tag your private functions with `@private`. Explanations of what the function does is not mandatory if it is straightforward (e.g. for your event handler, it's not necessary, so basically the tag `@private` is enough.

File: addons/web_editor/static/src/js/backend.js [hunk 6]
```diff
@@ -133,6 +131,8 @@ var FieldTextHtmlSimple = basic_fields.DebouncedField.extend({
         if (history) {
             history.reset();
         }
+        // currently we are unable to use the TranslatableFieldMixin in html widget
```
Comment (aab-odoo) [inline addons/web_editor/static/src/js/backend.js]: Actually you could, you simply need to override _renderTranslateButton() to render the correct template. Doing so, you don't need to re-defined the handler.

File: addons/web_editor/static/src/js/backend.js [hunk 7]
```diff
@@ -187,6 +187,18 @@ var FieldTextHtmlSimple = basic_fields.DebouncedField.extend({
         }
         return value;
     },
+
+    _renderTranslateButton: function () {
```
Comment (aab-odoo) [inline addons/web_editor/static/src/js/backend.js]: I would keep the same API as in the TranslatableMixin, i.e. this function returns a jQuery node, and it's the responsibility of the caller to insert it at the right place (i'm not saying that this is a better design, but for the sake of consistency, i would do that).

File: addons/web_editor/static/src/js/backend.js (hunk 5 above)
Comment (dka-odoo) [inline addons/web_editor/static/src/js/backend.js]: added this method to the handler section :)

File: addons/web_editor/static/src/js/backend.js (hunk 7 above)
Comment (dka-odoo) [inline addons/web_editor/static/src/js/backend.js]: alright! . keep the same consistency

File: addons/web_editor/static/src/js/backend.js (hunk 6 above)
Comment (dka-odoo) [inline addons/web_editor/static/src/js/backend.js]: need to return the TranslatableFieldMixin, after that it can be use here :)

**Comment by aab-odoo** [comment]
Hello,

Merged at 7ada921e67c77e0a9f9ad83f1f685e51c364171d. Finally, we decided not to handle the translate feature anymore in the html_frame widget as 1) it has been disabled in mass_mailing (see de403dc1ab440b9b08ea7fce189cd157745dfe74), and 2) it wasn't correctly working anyway (even before the new views).

Thanks for you work.
Regards

**Pull request closed without merging**